internal/component/prometheus/remote/queue/serialization/serializer.go

```diff
@@ -0,0 +1,223 @@
+	logger              log.Logger
+	self                actor.Actor
+	// Every 1 second we should check if we need to flush.
+	flushTestTimer *time.Ticker
+	series         []*types.TimeSeriesBinary
+	meta           []*types.TimeSeriesBinary
+	msgpBuffer     []byte
+	stats          func(stats types.SerializerStats)
+}
+
+func NewSerializer(cfg types.SerializerConfig, q types.FileStorage, stats func(stats types.SerializerStats), l log.Logger) (types.Serializer, error) {
+	s := &serializer{
+		maxItemsBeforeFlush: int(cfg.MaxSignalsInBatch),
+		flushFrequency:      cfg.FlushFrequency,
+		queue:               q,
+		series:              make([]*types.TimeSeriesBinary, 0),
+		logger:              l,
+		inbox:               actor.NewMailbox[*types.TimeSeriesBinary](),
+		metaInbox:           actor.NewMailbox[*types.TimeSeriesBinary](),
+		cfgInbox:            actor.NewMailbox[types.SerializerConfig](),
+		flushTestTimer:      time.NewTicker(1 * time.Second),
+		msgpBuffer:          make([]byte, 0),
+		lastFlush:           time.Now(),
+		stats:               stats,
+	}
+
+	return s, nil
+}
+func (s *serializer) Start() {
+	// All the actors and mailboxes need to start.
+	s.queue.Start()
+	s.self = actor.Combine(actor.New(s), s.inbox, s.metaInbox, s.cfgInbox).Build()
+	s.self.Start()
+}
+
+func (s *serializer) Stop() {
+	s.queue.Stop()
+	s.self.Stop()
+}
+
+func (s *serializer) SendSeries(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.inbox.Send(ctx, data)
+}
+
+func (s *serializer) SendMetadata(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.metaInbox.Send(ctx, data)
+}
+
+func (s *serializer) UpdateConfig(ctx context.Context, cfg types.SerializerConfig) error {
+	return s.cfgInbox.Send(ctx, cfg)
+}
+
+func (s *serializer) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Check for config which should have priority. Selector is random but since incoming
+	// series will always have a queue by explicitly checking the config here we always give it a chance.
+	// By pulling the config from the mailbox we ensure it does NOT need a mutex around access.
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case cfg, ok := <-s.cfgInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		s.maxItemsBeforeFlush = int(cfg.MaxSignalsInBatch)
+		s.flushFrequency = cfg.FlushFrequency
+		return actor.WorkerContinue
+	default:
+	}
+
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-s.inbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.Append(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case item, ok := <-s.metaInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.AppendMetadata(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append metadata to serializer", "err", err)
+		}
+		return actor.WorkerContinue
```

Comment:
Do we need these two cases? The inboxes are of the same type... the handling is almost the same... only error message is different. Maybe we can have just one inbox? I see that later in `store` you do split them, but maybe there's some way to simplify?

Reply:
Its possible, I think we can table this until the network PR. Metadata and time series have different write characteristics. Its a matter of where we want to split them. We could have one work stream and split them at the network level instead of splitting at the beginning like we do here.